### ISSUE TITLE
cached_result! fails to compile if no error is emitted from function with 'cannot infer type'

### ISSUE THREAD

**Issue by thorhs** [jaemk/cached#56]
Hi,

I've been banging my head against the wall for a few hours trying to compile my initial test of the cached crate.  I defined a cached_result!, but since I was just testing this out, I left out either the Err or the Ok case (basically empty function). This kept breaking the compile with 'cannot infer type'.  Now, a more seasoned rust developer might figure this out quickly, but alas... I did not.

After finally compiling with -Z macro-backtrace, I to gain a better unerstanding of the problem:
```
64  | / macro_rules! cached_result {
65  | |     // Unfortunately it's impossible to infer the cache type because it's not the function return type
66  | |     ($cachename:ident : $cachetype:ty = $cacheinstance:expr ;
67  | |      fn $name:ident ($($arg:ident : $argtype:ty),*) -> $ret:ty = $body:expr) => {
...   |
79  | |             let val = (||$body)()?;
    | |                       ^^^^^^^^^^^^ cannot infer type
...   |
84  | |     };
85  | | }
    | |_- in this expansion of `cached_result!` (#1)
    | 
   ::: src/quakes.rs:173:1
    |
173 | / cached_result! {
174 | |     HISTORICAL: UnboundCache<i64, String> = UnboundCache::new();
175 | |     fn get_historical_quake_week(week: i64) -> Result<String, String> = {
176 | |         Ok("Ok".to_owned())
177 | |     }
178 | | }
    | |_- in this macro invocation (#1)
```

I mostly wanted to get this out there in case others stumble upon the same error, it might warrant a mention in the documentation that the _result macro requires both arms of the Result to be implemented in the function.  Perhaps this could be fixed in the macro since the return type is supplied, but I'm no there yet myself.

**Comment by thorhs** [comment]
I decided to go ahead and try my hands at macros.  #57 does seem to fix the issue at hand, but feel free to improvise as needed.

**Comment by jaemk** [comment]
released in 0.19.0

**Comment by doivosevic** [comment]
can this then be closed?
